### ISSUE TITLE
Empty csrf-token

### ISSUE THREAD

**Issue by A5308Y** [NoRedInk/elm-rails#11]
Hi there!

Thanks for elm-rails! 

I'm having a some trouble with the csrf-tokens.

I followed https://blog.abevoelker.com/2016-09-05/adding-elm-to-a-rails-application using webpack to integrate elm in a rails app and added elm-rails.

I used
 
```
Rails.send "PATCH" url body
      |> Rails.fromJson (Rails.decoder success failure)
```

to build the request, but my Rails console complains with: "Can't verify CSRF token authenticity." I checked Rails.csrfToken in my view and it returns `Nothing` as well, although in my header I can see `<meta name="csrf-token" content="...">`

Disabling forgery_protection let's me see the sent data arrive in rails. So I guess everything except the csrf token is okay.

The documentation for `csrfToken` wisely pointed out that it only returns a token  

> If there was a `<meta name="csrf-token">` tag in the page's `<head>` *when elm-rails loaded*.

So I guess elm-rails is loaded too early?

I've added `<%= javascript_include_tag *webpack_asset_paths("application") %>` "right above the closing </body> tag" as suggested in the blog post above. 

I've tried explicitly enclosing `Elm...embed` in `$( window ).load` to no avail. My next idea was to try around with timeouts in the JS files, but that just seemed wrong.

So I'd be glad to get some help.

How do I find out if the load order is really the problem? If it is: How do I change it? From what I understand webpack runs as its own server, serving the elm code, which might be responsible for timing issues. Do I have to change something about webpack?

I've seen https://github.com/NoRedInk/elm-rails/commit/7c6ba931dc39ccfcd3cbe85e0809732e75647987 which looks like it could solve my problem, but I don't know how to use code that is not packaged yet. Is there something like bundler's source git?

Any other ideas?

Thanks in advance!

**Comment by eeue56** [comment]
@A5308Y can you share your elm-package.json? what version of elm-rails + elm are you using?

**Comment by A5308Y** [comment]
@eeue56 Ah yes. Sorry!

```

    "dependencies": {
        "NoRedInk/elm-rails": "4.0.5 <= v < 5.0.0",
        "elm-lang/core": "4.0.5 <= v < 5.0.0",
        "elm-lang/html": "1.1.0 <= v < 2.0.0"
    },
    "elm-version": "0.17.1 <= v < 0.18.0"
```

**Comment by eeue56** [comment]
So the problem is that with Elm 0.17, modules are no longer evaluated lazily. They're evaluated as soon as the JS is included - meaning that the wrapper for csrfToken is less reliable and more dependent on order. I'll have a fix in a minute

**Comment by A5308Y** [comment]
Oh wow! Cool! Thanks for the quick response!

**Comment by eeue56** [comment]
@A5308Y 

I was unable to reproduce your problem, but I believe you it's a problem. Can you check to see if this solution fixes it for you?
- open `elm-stuff/packages/NoRedInk/elm-rails/4.0.5/src/Native/Rails.js` and replace it with https://github.com/NoRedInk/elm-rails/blob/fix-rails-loading/src/Native/Rails.js
- open `elm-stuff/packages/NoRedInk/elm-rails/4.0.5/src/Rails.elm` and make any non-breaking modification, like add a comment or an empty line. This is to make sure the compiler realises the code needs to be recompiled
- Try what you were doing before

**Comment by A5308Y** [comment]
@eeue56 I tried your fix, but it didn't solve the problem for me. I could confirm that it's a loading problem though: I pasted your fix in my js-console and afterwards I tried the same with the code in the elm package. In both cases `_NoRedInk$elm_rails$Native_Rails.csrfToken` returned a token.

**Comment by eeue56** [comment]
It's hard to go anywhere else without some failing code to play around with. Could you share your Elm code where you actually trigger the HTTP request, if possible?

**Comment by A5308Y** [comment]
The following already returns "No csrfToken" every time.

``` javascript
import Rails

view : Model -> Html Msg
view model =
    div [] [ text (Maybe.withDefault "No csrfToken" Rails.csrfToken) ]
```

I think I made a mistake when recompiling though. The following:

``` javascript
var _NoRedInk$elm_rails$Native_Rails = function(){
  // when we aren't in an actual dom
  if ((typeof window === "undefined") || (typeof process === 'object')){
    return {
      csrfToken : { ctor : 'Just noWindow' }
    };
  }

  var csrfTokenNode = document.head.querySelector('meta[name="csrf-token"]');
  var csrfToken =
      (csrfTokenNode === null || (typeof csrfTokenNode.content !== "string"))
          ? { ctor: 'Just noContent' }
          : { ctor: 'Just', _0: csrfTokenNode.content };

  return {
    csrfToken: csrfToken
  }
}();
```

Still returns "No csrfToken". It should be either "noWindow" or "noContent", right?

**Comment by eeue56** [comment]
@A5308Y You can't set a `Just` value in native like that, it has to be `{ ctor: "Just", _0: "noWindow" }`

This explains how things get compiled -> https://github.com/NoRedInk/take-home/wiki/Native-types#union-types

That native code will suffer from the loading problem, if that is the cause. 

This should be the contents of that file:

``` javascript
var _NoRedInk$elm_rails$Native_Rails = function(){

  var getCsrfToken = function() {
    // when we aren't in an actual dom
    if ((typeof window === "undefined") || (typeof process === 'object')){
      return { ctor : 'Nothing' };
    }

    var csrfTokenNode = document.head.querySelector('meta[name="csrf-token"]');
    var csrfToken =
        (csrfTokenNode === null || (typeof csrfTokenNode.content !== "string"))
            ? { ctor: 'Nothing' }
            : { ctor: 'Just', _0: csrfTokenNode.content };
    return csrfToken;
  };

  return {
    get csrfToken () {
      return getCsrfToken();
    }
  }
}();
```

and it's vital that you trigger a recompile by changing some of the .elm files

**Comment by A5308Y** [comment]
Ok. I'll try to get this working and come back with more information for you. Thanks for your help so far!

**Comment by A5308Y** [comment]
@eeue56  

I'v verified that the code changes in Rails.js got picked up by webpack. 

I got 'process === object' from:

``` javascript
var _NoRedInk$elm_rails$Native_Rails = function(){

  var getCsrfToken = function() {
    // when we aren't in an actual dom
    if (typeof process === 'object') {
      return { ctor :  'Just', _0: 'process === object' };
    } else if (typeof window === "undefined") {
      return { ctor :  'Just', _0: 'Undefined Window' }
    };

    var csrfTokenNode = document.head.querySelector('meta[name="csrf-token"]');
    var csrfToken =
        (csrfTokenNode === null || (typeof csrfTokenNode.content !== "string"))
            ? { ctor: 'Just', _0: 'No Content' }
            : { ctor: 'Just', _0: csrfTokenNode.content };
    return csrfToken;
  };

  return {
    get csrfToken () {
      return getCsrfToken();
    }
  }
}();
```

But I have no idea what that means.

**Comment by eeue56** [comment]
Hmmm, okay. Remove the check for `process === 'object'` and does it work okay?

**Comment by A5308Y** [comment]
Yes! Great! I see a token! Thank you!

**Comment by eeue56** [comment]
Okidoki, I'll update the package in a bit once I've rethought the browser check

**Comment by A5308Y** [comment]
Cool! The request works like a charm as well now. Thanks again! I'll be waiting eagerly ;-)

**Comment by eeue56** [comment]
@A5308Y can you try this?

```
var _NoRedInk$elm_rails$Native_Rails = function(){
  var getCsrfToken = function() {
    // when we aren't in an actual dom, short curcuit
    // we can't always trust this test because mocking
    if (typeof window === "undefined"){
      return { ctor : 'Nothing' };
    }

    var csrfToken = { ctor: 'Nothing' };
    var csrfTokenNode = null;

    try {
      csrfTokenNode = document.head.querySelector('meta[name="csrf-token"]');
    } catch (e){
      // ignore document-based errors
    }

    if ((csrfTokenNode !== null) && (typeof csrfTokenNode.content === "string")){
      csrfToken = { ctor: 'Just', _0: csrfTokenNode.content };
    }

    return csrfToken;
  };

  return {
    get csrfToken () {
      return getCsrfToken();
    }
  }
}();

```

thanks!

**Comment by A5308Y** [comment]
Yep! Works for me.

👍

**Comment by eeue56** [comment]
@A5308Y published as 4.0.6, thanks for the report and testing the code!

**Comment by A5308Y** [comment]
Thanks for the fix and the incredible response time! Works for me out of the package now.
